### PR TITLE
Force automatic ndk download when building for Android

### LINKED ISSUE #155576: Decide on a path to stop stripping `libapp.so` and `libflutter.so` by default for android builds
Context https://github.com/flutter/flutter/issues/60240.

We currently strip debug symbols when [building the the AOT snapshot](https://github.com/flutter/flutter/blob/bd909542a33ab1d5249363a2434ae50ee468094f/packages/flutter_tools/lib/src/base/build.dart#L146) and when [building the engine](https://github.com/flutter/engine/blob/ba4d39ad3cda8f2a81425a3f1a4c4157d8380bc5/tools/gn#L1113). This means that when Flutter developers build an appbundle (`flutter build appbundle`) these debug symbols are not available to the Android Gradle plugin (AGP), and therefore they obviously can't be included in the appbundle, which results in the warning in the play store that we see in the linked issue.

Since AGP 4.1, AGP itself can handle both stripping the debug symbols from the native code, and automatically including it in the appbundle in a way that is available to app stores, but does not get delivered to end users* (see https://support.google.com/googleplay/android-developer/answer/9848633#upload_file). So ideally, we could simply stop stripping these files when building for Android, and instead let AGP handle the debug symbols.

HOWEVER - the Flutter Gradle plugin has for some time now vended a recommended version of the NDK (`flutter.ndkVersion`) that we use to specify the `ndkVersion` property in apps gradle files, and _importantly_, if an app specifies an ndk version, and doesn't have that ndk version installed locally, **the app will build but silently fail to strip the debug symbols**, meaning that they end up in the final app (greatly inflating the app size).

So before we can stop stripping the debug symbols from these files by default, we need to decide on a path that ensures we don't create a foot gun where developers build, unaware that they don't have the ndk version they are specifying installed, and end up with unnecessarily large apps.

In my mind there are two main options, but other suggestions are welcome:
1. Get rid of `flutter.ndkVersion` entirely via a migrator. If users don't specify an ndk version, apps will instead use the default version provided by AGP. This is option two in this [section of the android docs](https://developer.android.com/studio/projects/configure-agp-ndk#agp_version_41) - AGP will download the version by default. There is also a community issue [suggesting we do this](https://github.com/flutter/flutter/issues/139427).
2. Keep `flutter.ndkVersion` around, and instead modify the flutter tool/flutter doctor to ensure that we check/enforce that users have the version of the ndk that we specify installed.

\* More specifically, the symbols get stripped and included in a `BUNDLE-METADATA` subdirectory of the [app bundle format](https://developer.android.com/guide/app-bundle/app-bundle-format), see [this comment](https://github.com/flutter/engine/pull/50443#issuecomment-1932863940) for more details.

### PR DESCRIPTION
**Pull request by gmackall**
In preparation for changing engine builds to be unstripped by default https://github.com/flutter/engine/pull/52852, which will allow us to make progress towards resolving https://github.com/flutter/flutter/issues/60240.

Tricks AGP in to downloading the NDK when building a flutter app (that uses the FGP, which to my knowledge is all ways of building flutter apk/aab/aar).

I want to follow this up by modifying the tool to search for the log line that the NDK is missing (making it throw an error in that case) as a safeguard, because that would be the last line of defense before we accidentally build a bloated app. The safeguard won't work for add to app, from what I understand, because while they use the FGP (so they should be forced to be download the NDK) they don't invoke the flutter tool, and therefore won't invoke the custom error handling of [`gradle_errors.dart`](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/android/gradle_errors.dart)

Some details

1. Respects existing values for `externalNativeBuild.cmake.path` for apps that actually use it.
2. Silences some warnings that would otherwise appear for add to app builds or builds that manually invoke gradle:
```
C/C++: CMake Warning (dev) in CMakeLists.txt:
C/C++:   No project() command is present.  The top-level CMakeLists.txt file must
C/C++:   contain a literal, direct call to the project() command.  Add a line of
C/C++:   code such as
C/C++:     project(ProjectName)
C/C++:   near the top of the file, but after cmake_minimum_required().
C/C++:   CMake is pretending there is a "project(Project)" command on the first
C/C++:   line.
C/C++: This warning is for project developers.  Use -Wno-dev to suppress it.
C/C++: CMake Warning:
C/C++:   Manually-specified variables were not used by the project:
C/C++:     CMAKE_EXPORT_COMPILE_COMMANDS
C/C++:     CMAKE_LIBRARY_OUTPUT_DIRECTORY
C/C++:     CMAKE_RUNTIME_OUTPUT_DIRECTORY
```
3. Our ci installs the NDK at an abnormal place that AGP can't find without help. I've modified all the `build.gradle`s that we have checked in to point to the pre-installed path. **But some of our tests make a new app from the templates, and those tests will now start downloading the NDK** (as they won't be able to find it at it's current path from templates). We could resolve this by actually fixing https://github.com/flutter/flutter/issues/136666. This would be a very significant lift from what I understand - we rely on this hardcoding in a lot of places in our infra.

Fixes https://github.com/flutter/flutter/issues/155576

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by reidbaker** [comment]
Does this fix https://github.com/flutter/flutter/issues/155576?

**Comment by reidbaker** [comment]
@gmackall can you also link this pull request to the internal bug tracking allowing ndk download for symbol stripping?

File: dev/a11y_assessments/android/app/build.gradle [hunk 1]
```diff
@@ -35,7 +35,13 @@ if (keystorePropertiesFile.exists()) {
 android {
     namespace "com.example.a11y_assessments"
     compileSdk flutter.compileSdkVersion
-    ndkVersion flutter.ndkVersion
+
+    // Flutter's CI installs the NDK at a non-standard path.
```
Comment (reidbaker) [inline dev/a11y_assessments/android/app/build.gradle:39]: Include a link (go or direct) that points to where that system install is configured.

File: dev/benchmarks/multiple_flutters/android/app/build.gradle [hunk 2]
```diff
@@ -16,6 +16,13 @@ android {
     namespace "dev.flutter.multipleflutters"
     compileSdk 34
 
+    // Flutter's CI installs the NDK at a non-standard path.
+    String systemNdkPath = System.getenv("ANDROID_NDK_PATH")
+    if (systemNdkPath != null) {
+        ndkVersion = "26.3.11579264"
```
Comment (reidbaker) [inline dev/benchmarks/multiple_flutters/android/app/build.gradle:22]: Please include a comment indicating why this version and why not flutter.ndkVersion and when it should be updated?

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 3]
```diff
@@ -859,6 +860,34 @@ class FlutterPlugin implements Plugin<Project> {
...
+        }
+
+        // Otherwise, point to an empty CMakeLists.txt, and ignore associated warnings.
+        gradleProject.android {
+            externalNativeBuild {
+                cmake {
+                    // Respect the existing configuration if it exists - the NDK will already be
+                    // downloaded in this case.
+                    path = flutterSdkRootPath + "/packages/flutter_tools/gradle/src/main/groovy/CMakeLists.txt"
+                    //arguments "-Wno-dev"
```
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:877]: delete comment?

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 4]
```diff
@@ -859,6 +860,34 @@ class FlutterPlugin implements Plugin<Project> {
...
+                }
+            }
+
+            defaultConfig {
+                externalNativeBuild {
+                    cmake {
+                        arguments("-Wno-dev", "--no-warn-unused-cli")
```
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:884]: add a comment explaining why these flags and what would cause them to change.

Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:884]: Also if there is a way we can share these flags with some other code where we define similar flags lets consider that.

**Comment by gmackall** [comment]
> Does this fix #155576?

Yes, thanks - added

**Comment by gmackall** [comment]
> @gmackall can you also link this pull request to the internal bug tracking allowing ndk download for symbol stripping?

If you mean link the bug here, then it is b/152075797. But do you mean to put a link to the PR on the bug? Can do that too if desired

File: dev/a11y_assessments/android/app/build.gradle (hunk 1 above)
Comment (gmackall) [reply dev/a11y_assessments/android/app/build.gradle:39]: added

**Review by reidbaker — DISMISSED** [review]
.

**Review by reidbaker — COMMENTED** [review]
.

File: dev/a11y_assessments/android/app/build.gradle [hunk 5]
```diff
@@ -35,7 +35,15 @@ if (keystorePropertiesFile.exists()) {
...
+    // This non-standard structure is initially created by
+    // https://github.com/flutter/engine/blob/main/tools/android_sdk/create_cipd_packages.sh.
```
[the remark this answers is not in the document]
Comment (gmackall) [reply dev/a11y_assessments/android/app/build.gradle]: Changed to referencing the `3.27.0` tag

File: dev/benchmarks/multiple_flutters/android/app/build.gradle (hunk 2 above)
Comment (reidbaker) [reply dev/benchmarks/multiple_flutters/android/app/build.gradle:24]: Yes specifically there is still value in articulating when the value should be updated for all of the manually defined version comments.

**Review by reidbaker — APPROVED** [review]

File: .ci.yaml [hunk 6]
```diff
@@ -5503,7 +5503,7 @@ targets:
       add_recipes_cq: "true"
       dependencies: >-
         [
-          {"dependency": "android_sdk", "version": "version:33v6"},
+          {"dependency": "android_sdk", "version": "version:35v1"},
```
[the remark this answers is not in the document]
Comment (gmackall) [reply .ci.yaml:5506]: Even though the `vX` goes down this is actually an upgrade (33->35)

[the remark this answers is not in the document]
Comment (reidbaker) [reply .ci.yaml:5506]: 🤦 I was reviewing too quickly.

File: dev/benchmarks/multiple_flutters/android/app/build.gradle (hunk 2 above)
Comment (gmackall) [reply dev/benchmarks/multiple_flutters/android/app/build.gradle]: This is actually enforced now anyways (if the version is wrong the build will fail). But I added a comment indicating that it must exactly match the CIPD hosted ndk version that we are pulling.

**Comment by gmackall** [comment]
Reason for revert: Made the tree red due to some std out.

**Pull request merged**